### DISCUSSION TITLE
Suggestion: option for continuous generation

### DISCUSSION THREAD

**Discussion by ribawaja** [zero01101/openOutpaint#139] [Ideas]
Hi, I've a suggestion. I find that when I'm trying to inpaint something, I generate many images to find the one that fits nicely into my composition. So I'm switching back and forth between sifting through existing images, and generating more images for the end of the list. Often I'm stuck waiting for more images to generate, while the GPU was idle while I was sifting.
It would be nice if there were an option to continuously generate new images while I'm going through the list of possibilities, to reduce the chance that I have to wait on the GPU for new generations. There could be a button to stop this continuous generation, and a limit on the number of images that are generated ahead. For example if the generate-ahead limit is 50, and I've reached position 20 in the list, it would continue generating images to a length of 70.

I don't know javascript, but I've managed to get this mostly working by playing around with app\js\ui\tool\dream.js.
Do you think this is something that would be broadly useful enough to incorporate into the official repo? If not, I'll keep it forked for private use and try to keep my changes merged with yours.
Happy to share the changes I've made.

**Comment by ribawaja** [comment]
I've been playing around with the proposed feature, it seems to be working nicely. I originally added buttons on the toolbar to be able to start and stop continuous generation, but I later removed as it didn't seem necessary.
The only UI added is a checkbox to enable continuous generation and a numeric value to specify how many images to generate ahead (based on highest navigated position in the generated list). These two UI items are added to the context settings for dream and img2img.

I'll create a pull request for your consideration tomorrow, it's small, maybe 50 lines of code.

**Comment by ribawaja** [comment]
And separately I'm considering adding a way to mark items in the generated list, and a way to scroll through just the marked items.
For example, if I've generated 100 potential inpaintings for a malformed hand, and I've narrowed it down to three choices, and I want to be able to compare just those three out of the 100.

**Reply by seijihariki** [reply]
A full revamp of the dream choosing interface would be greatly appreciated!

**Comment by seijihariki** [comment]
Hello! That is a cool idea! I think that a setting in thee settings page could work: a text box where you put a 0+ number, and it represents how many images to keep pre-generated as your first cement suggests.

That way, people who don't want this may leave it as 0, but people who want some images always generated ahead can set it to any numbers, depending on their computer generation speed/usual speed of their scrolling.

Could you open a draft pull request to the testing branch for now? It becomes easier to manage and review existing changes.

**Comment by zero01101** [comment]
happy new year to all :) ~~agreed with the settings menu suggestion; while i initially thought this doesn't seem like something i'd personally use, now i'm curious to give it a shot haha~~

all of the above was patently wrong; this is way too fun to hide in a meta settings menu and i will personally use this regularly

well except for the happy new year part

that part was ok

**Comment by ribawaja** [comment]
Created PR https://github.com/zero01101/openOutpaint/pull/140

One minor issue with the new code is that the new slider value is read at the beginning of a generation, and isn't re-read until the user starts another generation (i.e. either accepts an image or rejects the batch). Ideally the slider would dynamically update the generate-ahead count, but I didn't see an easy way to do this. For one, any tool can be active at the time a batch completes and we need the slider value. 

I think this feature is great, but if you guys think it's too arcane and would cause confusion for the average user, happy to bury it in "advanced settings" somewhere so that the casual user doesn't see it.

**Comment by zero01101** [comment]
look maybe it's the champale talking but this is one of those "holy crap this is awesome" things i didn't realize i wanted until i used it, so imma merge it into main because it deserves to be there :)

**Reply by ribawaja** [reply]
That's awesome, thanks for accepting it

**Reply by zero01101** [reply]
seriously, thanks for writing and submitting it; my first read of the discussion was "well i don't understand why anyone would want that" but after trying it once it turned immediately into "oh ok i want this"

**Comment by zero01101** [comment]
>I don't know javascript

![image](https://user-images.githubusercontent.com/1649724/210196017-38956824-c08e-4301-a4bc-b40d83ee2cf0.png)

**Reply by zero01101** [reply]
so the champale is still talking, and after dicking around with the meta config menu, i think this _shouldn't_ live there; this feels like something that's going to be more regularly interacted with than a set-and-forget parameter

anyway again @ribawaja seriously this is way cool and i really like it and thank you for suggesting/writing/submitting it <3

**Reply by ribawaja** [reply]
Haha, I have other software development experience, but this is literally my first javascript attempt, I was puzzling over it and googling the difference between let/var, how async works, etc.  I hope you did due diligence on the code review because I don't want to break anything.

I'm really enjoying openOutpaint and may have other ideas as I continue to use it.

**Reply by zero01101** [reply]
> I hope you did due diligence on the code review

![lil bad pokerface](https://user-images.githubusercontent.com/1649724/210211494-beaf664f-7dc5-4369-a1e1-24a3a2a93d52.png)
